### PR TITLE
Point the old bootstrapping tutorial URL at new StatefulSet docs

### PR DESCRIPTION
**Pull request by janetkuo**
cc @bprashanth @erictune @foxish @kow3ns @enisoc @kubernetes/sig-apps

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1707)

### CODE REVIEW

File: docs/tutorials/stateful-application/statefulset_peers.yaml [hunk 1]
```diff
@@ -16,8 +16,8 @@ spec:
   selector:
```
Comment (kargakis) [inline docs/tutorials/stateful-application/statefulset_peers.yaml]: Shouldn't we fix terminationGracePeriod for these fixtures not to be zero?

**Comment by enisoc** [comment]
Is the main purpose of this tutorial to teach users how to run a VM-like workload? Or is it to demonstrate StatefulSet patterns? If it's the latter, I think there will be a lot of overlap with #1722 which uses the same patterns, but for MySQL.

**Comment by janetkuo** [comment]
@enisoc the main purpose of this tutorial is teaching users how Kubenetes + pods are different from VMs, based on http://kubernetes.io/docs/user-guide/petset/bootstrapping/, and deprecating the bootstrapping guide

**Comment by enisoc** [comment]
OK in that case I'll try to minimize overlap in my tutorial.

File: docs/tutorials/stateful-application/statefulset_peers.yaml (hunk 1 above)
Comment (janetkuo) [inline docs/tutorials/stateful-application/statefulset_peers.yaml]: Done

File: docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+- foxish
+- janetkuo
+- kow3ns
+- smarterclayton
+
+---
+
+{% capture overview %}
+
+This page helps you become familiar with the runtime initialization of [Stateful Sets](/docs/user-guide/petset).
```
Comment (smarterclayton) [inline docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md]: This is not necessarily new to this change, but the topic of this item is "VM-style workload" but nothing in the first 40 lines explains why that is.  A sentence in the overview or objectives that explains how the title relates to what is being shown would be good.

**Comment by steveperry-53** [comment]
I'm starting a writing review now.

**Comment by steveperry-53** [comment]
See [this commit](https://github.com/kubernetes/kubernetes.github.io/pull/1707/commits/e5c5c539f8a3db73e866e1227736a63da9d23842) for suggested changes.

**Comment by enisoc** [comment]
Reviewed 7 of 8 files at r1, 2 of 2 files at r2, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[_data/tutorials.yml, line 56 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0dSdpr_Ri_-AwYzOE:-KY0dSdpr_Ri_-AwYzOF:b-uy8hoo) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/_data/tutorials.yml#L56)):*
> ```YAML
> - title: Stateful Applications
>   section:
>   - title: Running a VM-style workload using Stateful Set
> ```

Merge StatefulSet into one word in title and URL.

---

*[docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md, line 29 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0dv9Lbre_xB_cUt2G:-KY0dv9Lbre_xB_cUt2H:bvzkv7h) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md#L29)):*
> ```Markdown
> * Learn about the runtime initialization of StatefulSets.
> 
> * Run a VM-style workload by using a StatefulSet and a Kubernetes a feature called
> ```

Remove extra "a" before "feature".

---

*[docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md, line 51 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0eE7eKqx3w5FUJc-d:-KY0eE7fJtRBNIPAcVZq:bhrgna6) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md#L51)):*
> ```Markdown
> * Your Kubernetes cluster must be running Kubernetes version 1.5 or later.
> 
> * Familiarize yourself with the terminology in
> ```

Add "the" at the end.

---

*[docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md, line 73 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0eaPIWLtGlx3FNcy6:-KY0eaPIWLtGlx3FNcy7:bifg9mt) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md#L73)):*
> ```Markdown
> 
> To further simulate a VM, a Pod needs a way to initialize its user environment
> when the the Pod starts. Also, tools like `kubectl exec` must not be allowed to
> ```

Double "the".

---

*[docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md, line 84 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0eiFKjIf62BuVlz7t:-KY0eiFKjIf62BuVlz7u:bni6om3) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md#L84)):*
> ```Markdown
> 
> Instead of guaranteeing that containers will not be restarted, Kubernetes provides
> a way to ensure that when a container stops and restarts, all of of the state
> ```

Double "of".

---

*[docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md, line 201 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0fK8r68marS2VkN3Y:-KY0fK8r68marS2VkN3Z:bhrgna6) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md#L201)):*
> ```Markdown
> In this portion of the tutorial, you create a StatefulSet of nginx servers, and
> you make one of them a master. The other nginx servers will proxy requests to
> the master. See
> ```

Add "the" at the end.

---

*[docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md, line 315 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0g7c_zUNYRqsLL9x7:-KY0g7c_zUNYRqsLL9x8:boer69w) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md#L315)):*
> ```Markdown
> * Write the config to a `hostPath` volume shared with the parent StatefulSet.
> 
> **Important:** In practice all Pods should query their peers for the current
> ```

We should elaborate on why this is important, especially since my replicated MySQL tutorial does the opposite. The difference is in this case you're allowing the master to change dynamically, whereas mine requires that the master is always index 0.

---

*[docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md, line 323 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0fv2W87QTEA_uz0lN:-KY0fv2W87QTEA_uz0lO:bax5ebn) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/vm-style-workload-using-stateful-set.md#L323)):*
> ```Markdown
> {% capture whatsnext %}
> 
> * You can deploy some example StatefulSets found
> ```

TODO: Add a link to Replicated Stateful Application tutorial.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707)*

**Comment by enisoc** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[docs/tutorials/stateful-application/statefulset_vm.yaml, line 36 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707#-KY0hQaMv7cOlvKuGGNM:-KY0hQaN6HqfeV7O0SsX:bbssbzp) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/e5c5c539f8a3db73e866e1227736a63da9d23842/docs/tutorials/stateful-application/statefulset_vm.yaml#L36)):*
> ```YAML
>                     "/bin/sh",
>                     "-c",
>                     "for d in usr lib etc; do cp -vnpr /$d/* /${d}mnt; done;"
> ```

The overall script should somehow exit with a non-zero error code if the whole process didn't succeed. You don't want to run the real workload if the initialization was incomplete. In Bash you can just add `set -e` at the beginning of the script. I'm not sure if that's guaranteed to work in whatever `/bin/sh` refers to.

In general, this approach of trying to merge in potential changes from the image (with `cp --no-clobber`) after the PVC has been initialized once before is dangerous. Suppose you let everything initialize, then delete a `conf.d` file that contains default login credentials. A subsequent Pod restart will restore that file, creating a vulnerability. This example also doesn't account for /var being lost between Pod reschedulings, which can cause data loss and other unexpected behavior.

At the least, we should warn that you shouldn't actually do this; it's just a toy example to demonstrate some concepts. A step further would be to do nothing at all if the PVC has ever been initialized before. That would at least be safer, although it means you should never change the image.

Personally my recommendation is to drop this tutorial altogether for now. If there are still concepts we want to teach that aren't covered in other tutorials, we should start over and rethink how to present them in a less misleading way. I suspect I'm in the minority in that so I'll defer to the wisdom of the team.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1707)*

**Comment by janetkuo** [comment]
> from @enisoc: Personally my recommendation is to drop this tutorial altogether for now.

@erictune @foxish @kow3ns thoughts?

**Comment by enisoc** [comment]
To summarize: we discussed this offline and agreed to remove the VM-style tutorial for now because our other new StatefulSet tutorials cover the important concepts with use cases that have clearer motivations.

This PR will be repurposed to just point the old bootstrapping tutorial URL at new StatefulSet docs.

**Comment by janetkuo** [comment]
PR & title updated, PTAL

**Comment by foxish** [comment]
LGTM

**Comment by devin-donnelly** [comment]
I think we should absolutely add this note at the top of the documentation, but leave it where it is for now--it still applies to 1.4 users.

File: docs/user-guide/petset/bootstrapping/index.md [hunk 3]
```diff
@@ -9,6 +9,7 @@ assignees:
 - smarterclayton
 ---
 
-* Starting in Kubernetes version 1.5, PetSet has been renamed to [StatefulSet](/docs/concepts/abstractions/controllers/statefulsets).
-* This document has been deprecated. For information on working with StatefulSet, see the tutorial on [how to run replicated stateful applications](/docs/tutorials/stateful-application/run-replicated-stateful-application).
+* Starting in Kubernetes version 1.5, PetSet has been renamed to [StatefulSet](/docs/concepts/abstractions/controllers/statefulsets). To use (or continue to use) PetSet in Kubernetes 1.5, you _must_ migrate your exsiting PetSets to StatefulSets. For information on working with StatefulSet, see the tutorial on [how to run replicated stateful applications](/docs/tutorials/stateful-application/run-replicated-stateful-application).
+
+* __This document has been deprecated__, but can still apply if you're using Kubernetes version 1.4 or earlier.
```
Comment (steveperry-53) [inline docs/user-guide/petset/bootstrapping/index.md]: If we're going to say this document is deprecated but can still apply, then we need to keep the content of the document.

**Comment by enisoc** [comment]
@devin-donnelly I think it would make sense to keep the doc if it was helping 1.4 users to maintain existing PetSets (like if it was a Task page describing how to do a manual rolling update), but to me the point of this particular doc is encouraging users to adopt PetSet for new use cases. Since PetSet is deprecated, our plan was to discourage against creating new PetSets and instead recommend upgrading to 1.5+ and using StatefulSets.

**Comment by devin-donnelly** [comment]
@enisoc I understand. The only users I feel like we might miss are those that don't want to risk upgrading off 1.4 but still want to use PetSet (despite it being Alpha). I don't know if there are a lot of those users out there; any insight?

File: docs/user-guide/petset/bootstrapping/index.md (hunk 3 above)
Comment (devin-donnelly) [inline docs/user-guide/petset/bootstrapping/index.md]: Yep. I'm suggesting below that we do _not_ actually delete the content.

**Comment by devin-donnelly** [comment]
@enisoc After a close reading, I strike my objection. Is there existing documentation that's a more basic "PetSet Guide" that needs to be marked as deprecated, however?

**Comment by steveperry-53** [comment]
@janetkuo Would you like to squash commits before we merge this?

**Comment by enisoc** [comment]
@devin-donnelly Yeah I think [user-guide/petset](https://github.com/kubernetes/kubernetes.github.io/blob/release-1.5/docs/user-guide/petset.md) fits your description for the basic guide that should probably be kept with just a deprecation warning at the top.

File: docs/user-guide/petset/bootstrapping/index.md [hunk 4]
```diff
@@ -1,243 +1,15 @@
...
+- erictune
+- foxish
+- janetkuo
+- kow3ns
+- smarterclayton
 ---
 
-* TOC
-{:toc}
+Starting in Kubernetes version 1.5, PetSet has been renamed to [StatefulSet](/docs/concepts/abstractions/controllers/statefulsets). To use (or continue to use) PetSet in Kubernetes 1.5, you _must_ migrate your exsiting PetSets to StatefulSets. For information on working with StatefulSet, see the tutorial on [how to run replicated stateful applications](/docs/tutorials/stateful-application/run-replicated-stateful-application).
```
Comment (enisoc) [inline docs/user-guide/petset/bootstrapping/index.md]: And should we link "migrate ..." directly to the PetSet upgrade task?

**Comment by enisoc** [comment]
I didn't see a PR yet for the above so created https://github.com/kubernetes/kubernetes.github.io/pull/1874

**Comment by devin-donnelly** [comment]
I'll go ahead and merge this.

**Pull request merged**